docs/src/proposals/traced-transaction-construction.md

```diff
@@ -0,0 +1,247 @@
+tx's account keys.
+
+While MessageProcessor's `verify_accounts` is enforced to these accessed
+accounts like normally-accessed accounts, this syscall can bypass the CPI
+boundaries by always accessing to the outermost execution environment (the
+whole AccountsDB account set while simulated, all of account keys while
+executed). This bypass semantic is needed to work with existing CPI usage.
+
+Then, otherwise noted, the returned `AccountInfo` should behave in the exactly
+same way both in the simulated and executed mode. This is desired for
+determisnism while developing and is mandated for security to lightly protect
+users from bad programs pretending to be harmless only in simulation.
+
+
+### example: the associated token account program (demonstrating the problem #1)
+
+```patch
+diff --git a/associated-token-account/program/src/processor.rs b/associated-token-account/program/src/processor.rs
+index 98eb08b..fa8ae50 100644
+--- a/associated-token-account/program/src/processor.rs
++++ b/associated-token-account/program/src/processor.rs
+@@ -22,13 +22,12 @@ pub fn process_instruction(
+     let account_info_iter = &mut accounts.iter();
+ 
+     let funder_info = next_account_info(account_info_iter)?;
+-    let associated_token_account_info = next_account_info(account_info_iter)?;
+     let wallet_account_info = next_account_info(account_info_iter)?;
+     let spl_token_mint_info = next_account_info(account_info_iter)?;
+-    let system_program_info = next_account_info(account_info_iter)?;
+-    let spl_token_program_info = next_account_info(account_info_iter)?;
++    let system_program_info = access_as_unsigned(funder_info.owner(), READ)?;
++    let spl_token_program_info = access_as_unsigned(spl_token_mint_info.owner(), READ)?;
+     let spl_token_program_id = spl_token_program_info.key;
+-    let rent_sysvar_info = next_account_info(account_info_iter)?;
++    let rent_sysvar_info = access_as_unsigned(sysvars::rent::id(), READ)?;
+ 
+     let (associated_token_address, bump_seed) = get_associated_token_address_and_bump_seed_internal(
+         &wallet_account_info.key,
+@@ -36,10 +35,7 @@ pub fn process_instruction(
+         program_id,
+         &spl_token_program_id,
+     );
+-    if associated_token_address != *associated_token_account_info.key {
+-        msg!("Error: Associated address does not match seed derivation");
+-        return Err(ProgramError::InvalidSeeds);
+-    }
++    let associated_token_account_info = access_as_unsigned(associated_token_address, READ_WRITE)?;
+ 
+     let associated_token_account_signer_seeds: &[&[_]] = &[
+         &wallet_account_info.key.to_bytes(),
+```
+
+This reduces the number of required accounts for CPI and clients from 7 to 3, reducing by 4.
+(Obviously, we can't actually deploy these changes because this will break
+compatibility. This is only for illustration of effective applicability to
+moderately-complex program construct.)
```

Comment:
also, maybe worth to mention that this doesn't reduce the resultant tx binary size, we need still this badly for more composability: https://github.com/solana-labs/solana/pull/15927 cc: @jstarry 

The good thing is that this nor the big tx block each other. :)